pallets/xcm-core-buyer/src/lib.rs

```diff
@@ -0,0 +1,494 @@
+// Copyright (C) Moondance Labs Ltd.
+// This file is part of Tanssi.
+
+// Tanssi is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Tanssi is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Tanssi.  If not, see <http://www.gnu.org/licenses/>
+
+//! # XCM Core Buyer Pallet
+//!
+//! This pallet allows collators to buy parathread cores on demand.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(any(test, feature = "runtime-benchmarks"))]
+mod benchmarks;
+pub mod weights;
+
+use {
+    crate::weights::WeightInfo,
+    dp_core::ParaId,
+    frame_support::{
+        pallet_prelude::*,
+        traits::fungible::{Balanced, Inspect},
+    },
+    frame_system::pallet_prelude::*,
+    parity_scale_codec::EncodeLike,
+    sp_runtime::traits::{AccountIdConversion, Convert, Get},
+    sp_std::vec,
+    sp_std::vec::Vec,
+    staging_xcm::prelude::*,
+    staging_xcm::v3::{InteriorMultiLocation, MultiAsset, MultiAssets, Xcm},
+    tp_traits::ParathreadParams,
+};
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+    #[pallet::pallet]
+    #[pallet::without_storage_info]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// Overarching event type.
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+        type Currency: Inspect<Self::AccountId> + Balanced<Self::AccountId>;
+
+        type XcmSender: SendXcm;
+        /// Get encoded call to buy a core in the relay chain. This will be passed to the XCM
+        /// `Transact` instruction.
+        type GetPurchaseCoreCall: GetPurchaseCoreCall<Self::RelayChain>;
+        /// Get current block number, used in `validate_unsigned`.
+        type GetBlockNumber: Get<u32>;
+        /// How to convert a `ParaId` into an `AccountId32`. Used to derive the parathread tank
+        /// account in `interior_multilocation`.
+        type GetParathreadAccountId: Convert<ParaId, [u8; 32]>;
```

Comment:
Question. Who owns this account? Is it a chain identified by the corresponding `para_id`?

Reply:
yes exactly, it's a chain-derived account based on the paraId.

Reply:
for now this address is only usable for this purpose. However, in the future we might want the container-chain manager to have some power over this